### ISSUE TITLE
Milestone of CUDA & Triton

### ISSUE THREAD

**Issue by Albresky** [Albresky/leetgpu#1]
# Roadmap (CUDA & Triton)

## 目标

1. **掌握核心算子**：GEMM (FP16/INT8), FlashAttention, Softmax, LayerNorm.
2. **吃透硬件特性**：Memory Hierarchy (Shared Mem, Registers), Tensor Cores (WMMA/MMA), Hopper TMA.
3. **工程化能力**：Profiling (NCU), Benchmarking (Roofline), Triton DSL.

---

## Day 1: 内存模型与并行 (Memory Hierarchy & Tiling)

**目标**：理解 GPU 内存层级，解决 Bandwidth Bound 问题。
**硬件**：RTX 4090 (Ada Lovelace)

### 知识点

* **Memory Coalescing**: 为什么 `Row-Major` 和 `Col-Major` 读写性能差异巨大。
* **Shared Memory & Bank Conflicts**: 如何通过 Padding 解决 Bank Conflict。
* **Warp Divergence**: 避免 Warp 内的分支预测失败。

### LeetGPU

- [x] 1. **[Easy] Matrix Transpose**:
  * *Challenge*: 使用 Shared Memory 进行 Tiling，解决非合并访存问题。处理 Bank Conflicts (如 `input[i][j]` 映射到 Shared Mem `smem[j][i]` 时加 padding `smem[32][33]`)。

- [x] 2. **[Medium] Reduction**:
  * *Challenge*: 必须使用 `__shfl_down_sync` (Warp Shuffle 指令) 进行最后 32 个线程的规约，这是现代 CUDA 优化的标准操作，比 Shared Mem 更快。

- [x] 3. **[Easy] Vector Addition**:
  * *Challenge*: 用 `float4` 向量化读写（Vectorized Memory Access）来榨干带宽。

- [x] 4. **[Easy] Matrix Multiplication**:
  * *Challenge*: Matrix tiling, 高计算访存比，WMMA指令初探。

### 延伸

* 使用 `nsight-compute (ncu)` 分析 Transpose kernel，查看 **DRAM Throughput** 和 **L2 Cache Hit Rate**。

---

## Day 2: Tensor Core 与 GEMM 基础 (Compute Bound)

**目标**：理解矩阵乘法的分块优化，入门 Tensor Core。
**硬件**：RTX 4090 / H100

### 知识点

* **Hierarchical Tiling**: Block Tile -> Warp Tile -> Thread Tile。
* **Tensor Cores (WMMA API)**: 使用 `nvcuda::wmma` 进行 FP16 矩阵乘。
* **Data Layout**: `NHWC` vs `NCHW` 在 Tensor Core 下的优劣。

### LeetGPU

- [ ] 1. **[Medium] General Matrix Multiplication (GEMM)** (重点):
  * *Step 1*: 写一个 Naive 版本。
  * *Step 2*: 写一个 Shared Memory Tiling 版本。
  * *Step 3*: **强制使用 Tensor Core**。尝试用 `wmma::load`, `wmma::mma_sync`, `wmma::store` 实现。

- [ ] 2. **[Medium] INT8 Quantized MatMul**:
  * *Challenge*: 实现 `IMMA` (Integer Matrix Multiply Accumulate)。这是推理加速（Inference）的核心。注意处理 `scale` 和 `zero_point` 的量化逻辑。

### 阅读

* NVIDIA Blog: *Programming Tensor Cores in CUDA 9*.
* CUTLASS 文档中关于 Layout 和 Iterator 的概念。

---

## Day 3: Triton 与 Softmax

**目标**：从 CUDA 的繁琐细节转向 Triton 的块状编程思维。
**环境**：本地 RTX 4090

### 知识点

* **Triton Programming Model**: `@triton.jit`, `tl.load`, `tl.store`。
* **Block Pointers (New in Triton)**: 自动处理边界检查和 stride。
* **Fusion**: 算子融合的基本原理（减少 HBM 访问）。

### LeetGPU (用 Triton 求解) / 本地练习

- [ ] 1. **[Medium] Softmax**:
  * *CUDA 思路*: 需要多次 Global Memory 交互（求 Max -> 求 Sum -> Normalize）。
  * *Triton 思路*: 加载一行数据到 SRAM，在寄存器中完成 Max/Sum/Exp 计算，写回一次。
  * *关键点*: 数值稳定性（Minus Max）。

- [ ] 2. **[Medium] RMS Normalization**:
  * LLM 中常用的 Norm 层，适合用 Triton 实现，代码量极少但性能极高。

---

## Day 4: FlashAttention 前置技能 (Attention is all you need)

**目标**：手撸 Attention 机制，理解 IO 复杂度优化。
**硬件**：H100 (利用其庞大 SRAM)

### 知识点

* **Standard Attention**:  内存复杂度痛点。
* **Online Softmax**: 如何在不知道全局 Sum 的情况下计算 Softmax（FlashAttention 的数学基础）。

### LeetGPU

- [ ] 1. **[Medium] Softmax Attention**:
  * *Task*: 实现 , Scale, Mask, Softmax,  的完整流程。
  * *Challenge*: 尝试将 Kernel 分裂成两个（Gemm1 -> Softmax -> Gemm2）观察性能，思考为什么这很慢。

- [ ] 2. **[Hard] Causal Self-Attention**:
  * *Task*: 加入 Masking。在 GEMM 累加过程中直接将 Mask 位置设为 。

---

## Day 5: 挑战 FlashAttention (The Holy Grail)

**目标**：在 Triton/CUDA 中实现 FlashAttention V1/V2 逻辑。
**硬件**：RTX 4090 / H100

### 知识点

* **FlashAttention V2 Algorithm**:
* Tiling Q, K, V。
* Outer loop over blocks of K/V, Inner loop over blocks of Q (为了并行化)。
* **Recomputation**: 在 Backward 阶段重算 Attention score 而不存 Global Memory（虽然目前只做 Forward，但需理解原理）。

### 本地或 LeetGPU Hard 题改写

* **目标**: 针对 LeetGPU **[Hard] Multi-Head Attention** 或 **[Medium] Attention with Linear Biases**。
* **实现**: 不要用三个分开的 kernel。尝试写成一个 **Fused Kernel**。
  * 使用 Shared Memory 缓存、加载一块 。
  * 计算局部 ，更新局部 Max 和 Sum，Rescale 之前的 Accumulator。
  * AI Infra 终极考题。

---

## Day 6: 新硬件特性 (H100/Hopper & Async)

**目标**：利用 H100 特性（TMA, WGMMA）。
**硬件**：**LeetGPU H100 / B200 Instance**

### 知识点

* **Asynchronous Copy (cp.async)**: Ampere 引入，Hopper 增强。将 Global -> Shared 的数据搬运与计算重叠（Hiding Latency）。
* **TMA (Tensor Memory Accelerator)**: Hopper 专属，由硬件引擎负责搬运 Tensor，解放 SM。
* **Warpgroup (wgmma)**: 128 个线程协同完成矩阵乘。

### 练习方向

* 回到 **[Medium] GEMM**。
* **Level Up**: 尝试在 CUDA 中使用 `__pipeline_memcpy_async` (CUDA 12+)。
* 接触 H100 的 PTX 或 CUTLASS 3.x，尝试理解 **TMA Descriptor** 的构建。
* *加分项*： 解释为什么 H100 上 TMA 能显著提升 GEMM 效率（减少寄存器压力，减少指令数）。

---

## Day 7: 编译集成与 Benchmark 方法论

**目标**：像 Compiler Engineer 一样思考，不仅仅是写 Kernel。

### 1. Benchmark 标准流程 (Markdown Spec)

提交代码时，应包含以下 Benchmark 报告：

```markdown
# Kernel Performance Benchmark

## Environment
- **GPU**: NVIDIA H100 PCIe (80GB)
- **CUDA**: 12.1
- **Driver**: 535.xx
- **Framework**: PyTorch 2.1 (for baseline comparison)

## Methodology
1. **Warmup**: 10 iterations to settle clock frequencies.
2. **Measurement**: 100 iterations, taking median latency.
3. **Metric**: TFLOPS (Tera Floating-point Operations Per Second) and %SOL (Speed of Light).

## Results: FP16 GEMM (M=4096, N=4096, K=4096)

| Implementation | Latency (ms) | TFLOPS | Speedup vs CuBLAS |
| :--- | :--- | :--- | :--- |
| PyTorch (CuBLAS) | 0.15 | 900 | 1.0x |
| Naive CUDA | 12.0 | 11 | 0.01x |
| Tiled CUDA | 1.5 | 90 | 0.1x |
| **My WMMA Kernel** | **0.18** | **750** | **0.83x** |

## Roofline Analysis
- Compute usage: 85% of theoretical max.
- Memory Bandwidth: 40% (Compute bound kernel).

```

### 2. 怎么测 (How to Benchmark)

* **Python (Triton/PyTorch)**:
```python
import triton.testing
@triton.testing.perf_report(...)
def benchmark(M, N, K):
    # ... run kernel ...
    return ms

```

* **Roofline Model**: 确定该算子是 Compute Bound 还是 Memory Bound。
* *Arithmetic Intensity (AI)* = FLOPs / Bytes Transferred.
* H100 SXM: FP16 Tensor Core 算力 ~989 TFLOPS, 带宽 ~3.35 TB/s. 拐点极高，绝大多数算子如果不融合都是 Memory Bound。

---

## 总结提升

1. **LeetGPU 的 B200 / H200**:
* 这些卡支持 **FP8** 甚至 **FP4**。
* 在做 **[Medium] INT8 Quantized MatMul** 时，思考如果硬件支持 `e4m3` 或 `e5m2` (FP8) 数据类型，Kernel 需要怎么改？（提示：数据加载位数减半，带宽翻倍，WMMA 形状变化）。

2. **LLVM 结合**:
* 思考：如果在写一个类似 Triton 的编译器，如何自动生成 Day 1 中提到的 `padding` 来避免 Bank Conflicts？这叫 **Shared Memory Allocation Analysis**。

3. **“如何优化 GEMM？”**
* 不要只回答 Tiling。
* 而要回答：**Global Mem Coalescing -> Shared Mem Tiling (避免 Bank Conflict) -> Register Tiling -> Double Buffering (Prefetch) -> Tensor Core -> Async Copy (TMA)**。回答问题的深度就在此 :grin:。
